Implemented content-fairness ranking, resolved issue in novelty, cleaned up code and repo

jannikjw opened the pull request:
- Implemented content fairness ranking
- Changed novelty calculation (based on probabilities and on full dataset not only slate)
- extended experiments.py with more optional arguments
- changed user pairing to be based on user clusters
- deleted old measurement files
- fixed issues in cosine similarity (value below zero would mess up novelty metric -> all prediction scores are shifted to be above 0)

jannikjw committed:
experiments in prob ranking

jannikjw committed:
- attemps at prob-based-rec

jannikjw committed:
Added content_fairness scoring_function (not tested)

jannikjw committed:
change default experiments config

jannikjw committed:
Cleaned up notebooks and fixed cosine similarity

- moved analysis code from trecs_exp.ipynb to new notebook analysis.ipynb
- fixed issue in cosine similarity where scores would be negative and the novelty would be positive by adding
- deleted unnecessary files

jannikjw committed:
Fixed issue in novelty

- novelty relies on scores -> instead of absolute values, convert scores to probabilities to ensure similar scaling
- fixed bug in cosine_similarity()

jannikjw committed:
Fix errors

jannikjw committed:
Implemented user pairing based on clusters

Added the decision to pair users only with others who are not in the same cluster.
This affects recommendation similarity and interaction similarity measurement.
-> Should be turned into passable arguments for experiments.py

jannikjw committed:
Added recall metric

jannikjw committed:
added recall metric

madisonthantu commented on src/scoring_functions.py:
```diff
@@ -28,4 +33,40 @@ def entropy(predicted_user_profiles, predicted_item_attributes):
+
+
+def content_fairness(predicted_user_profiles, predicted_item_attributes):
+    slate_size = 10
```
if there's any chance that we reduce/change the slate size maybe consider making this a hyperparameter

jannikjw replied:
Unfortunately, you cannot pass parameters to the score_fn. It is obviously bad practice but I would say it's totally fine for our project.

madisonthantu commented on src/scoring_functions.py:
```diff
@@ -28,4 +33,40 @@ def entropy(predicted_user_profiles, predicted_item_attributes):
+    return re_ranked_scores
+
+
+def content_fairness(predicted_user_profiles, predicted_item_attributes):
```
I'm having trouble following this 😎

jannikjw replied:
Fair, I added documentation - does it make more sense now?

madisonthantu commented on wrapper/metrics/evaluation_metrics.py:
```diff
@@ -240,4 +245,59 @@ def measure(self, recommender):
+                Model that inherits from
+                :class:`~models.recommender.BaseRecommender`.
+        """
+        if self.k >= recommender.num_items_per_iter:
```
move to init

jannikjw replied:
the recommender is not know in init

madisonthantu commented on wrapper/metrics/evaluation_metrics.py:
```diff
@@ -240,4 +245,59 @@ def measure(self, recommender):
+        else:
+            shown_item_scores = np.take(recommender.predicted_scores.value, recommender.items_shown)
+            shown_item_ranks = np.argsort(shown_item_scores, axis=1)
+            top_k_items = np.take(recommender.items_shown, shown_item_ranks[:, self.k :])
```
not sure but should it be `shown_item_ranks[:,recommender.items_per_iter-self.k:]`

madisonthantu commented on clustering_experiments.ipynb:
```diff
@@ -0,0 +1,370 @@
+    },
+    {
+     "data": {
+      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAjoAAAGPCAYAAABVp3qbAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/NK7nSAAAACXBIWXMAAA9hAAAPYQGoP6dpAAEAAElEQVR4nOydd5hkZZm375Mq5+ru6jw558jAzJARxAAIKAoiIqKIurp+LuDuumZdddlFMYuoYEBJIsKAhCHDJCbn0DOdu6u6cjzp+6N6aqamuyeTz31dXBd9znve81ZN96lfPe/z/B7BNE0TCwsLCwsLC4u3IeIbvQALCwsLCwsLi9cKS+hYWFhYWFhYvG2xhI6FhYWFhYXF2xZL6Fh ...
```
no output shown for this cell onwards

madisonthantu reviewed (commented):
Left some comments but overall looks ~~beautiful~~

jannikjw committed:
Documented content_fairness

jannikjw committed:
Resolve review of PR

- fixed recall
- added provisions to scoring functions that divide by zero in model initialization
- extended experiments.py to configure user parameters

jannikjw committed:
Saving and changing default drift

jannikjw committed:
Re-implemented content fairness ranking

jannikjw committed:
Improved experiments.py

- fixed default value for user attention (should be negative)
- formatted argument defaults in general

jannikjw committed:
Merge branch 'main' into prob_based_rec

jannikjw commented:
I am right, you are wrong
